docarray/array/storage/base/getsetdel.py

```diff
@@ -76,35 +84,74 @@ def _del_docs_by_slice(self, _slice: slice):
         Override this function if there is a more efficient logic
         :param _slice: the slice used for indexing
         """
-        for j in range(len(self))[_slice]:
-            self._del_doc_by_offset(j)
+        ids = self._offset2ids.get_id(_slice)
+        self._del_docs(ids)
 
     def _del_docs_by_mask(self, mask: Sequence[bool]):
         """This function is derived and may not have the most efficient implementation.
         Override this function if there is a more efficient logic
         :param mask: the boolean mask used for indexing
         """
-        for idx, m in enumerate(mask):
-            if not m:
-                self._del_doc_by_offset(idx)
+        # TODO: rebuild offset2ids once
+        ids = list(itertools.compress(self._offset2ids, (not _i for _i in mask)))
+        self._del_docs(ids)
 
     def _del_all_docs(self):
         """This function is derived and may not have the most efficient implementation.
 
         Override this function if there is a more efficient logic"""
-        for j in range(len(self)):
-            self._del_doc_by_offset(j)
+        self._clear_storage()
+        self._offset2ids = Offset2ID()
 
-    # Setitem API
+    def _del_docs_by_ids(self, ids):
+        """This function is derived from :meth:`_del_doc_by_id`
+        Override this function if there is a more efficient logic
+
+        :param ids: the ids used for indexing
+        """
+        for _id in ids:
+            self._del_doc_by_id(_id)
+
+    def _del_docs(self, ids):
+        self._del_docs_by_ids(ids)
+        self._offset2ids.delete_by_ids(ids)
 
     @abstractmethod
-    def _set_doc_by_offset(self, offset: int, value: 'Document'):
+    def _clear_storage(self):
         ...
 
+    # Setitem API
+
+    def _set_doc_by_offset(self, offset: int, value: 'Document'):
+        self._set_doc(self._offset2ids.get_id(offset), value)
+
+    def _set_doc(self, _id: str, value: 'Document'):
+        if _id != value.id:
```

Comment:
Here we should think what happens if the `set_doc_by_id` is not successful.

Maybe we should update offset after and if no exception is raised